src/plugins/intel_gpu/src/runtime/ocl/ocl_common.hpp

```diff
@@ -20,7 +20,49 @@ class ocl_error : public ov::Exception {
     explicit ocl_error(cl::Error const& err);
 };
 
-#define OCL_ERR_MSG_FMT(err) ("[GPU] " + std::string(err.what()) + std::string(", error code: ") + std::to_string(err.err()))
+#define OCL_ERR_MSG_FMT(error) ("[GPU] " + std::string(error.what()) + std::string(", error code: ") + std::to_string(error.err()))
+
+
+/// WA: Force exit. Any opencl api call can be hang after CL_OUT_OF_RESOURCES.
+inline void force_exit() {
+    std::cerr << "[GPU] force exit.\n"
+              << "\tDue to the driver bug any subsequent OpenCL API call will cause application hang, "
+              << "so GPU plugin can't finish correctly.\n"
+              << "\tPlease try to update the driver or reduce memory consumption "
+              << "(use smaller batch size, less streams, lower precision, etc)"
+              << "to avoid CL_OUT_OF_RESOURCES exception" << std::endl;
+    std::_Exit(-1);
+}
+
+inline bool is_device_available(const device_info& info) {
+    ocl_device_detector detector;
+    auto devices = detector.get_available_devices(nullptr, nullptr);
+    for (auto& device : devices) {
+        if (device.second->get_info().uuid.uuid == info.uuid.uuid) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+inline void rethrow_or_exit(std::string message, cl_int error, const device_info& info) {
+    if (error != CL_OUT_OF_RESOURCES) {
+        OPENVINO_THROW(message);
+    }
+    // For CL_OUT_OF_RESOURCES exception there are 2 possible cases:
+    // 1. Real out of resource which means that plugin must exit
+    // 2. Device is lost during application run, plugin may throw an exception
+    if (is_device_available(info)) {
+        OPENVINO_THROW(message);
+    } else {
+        force_exit();
```

Comment:
A bit confused by the name of the change and the change itself
Here, if the device is lost, `force_exit()` is also called

May also change error message to a more specific when device is not available